File: Secant/Secant.jl
using BenchmarkTools
f(x) = x^2 - 11
function secant_method(f,a, b, tol)
    if f(a)*f(b) > 0
        println("No root found.")
    else
        while (b - a) >= tol
            midpoint = a - f(a)*(b-a)/(f(b)-f(a))
            if f(midpoint) == 0
                return midpoint
            elseif f(a)*f(midpoint) < 0
                b = midpoint
            else
                a = midpoint
            end
        end       
        return a - f(a)*(b-a)/(f(b)-f(a))
    end
end

result = @benchmark secant_method(f,-1, 5, 0.0001)

println("Answer: ", secant_method(f,-1, 5, 0.0001))
println("Minimum time: $(minimum(result.times) / 1e9) seconds")
println("Maximum time: $(maximum(result.times) / 1e9) seconds")
println("Average time: $(mean(result.times) / 1e9) seconds")

File: Secant/Secant.py
import timeit

fx = lambda x: x**2 - 11

def secant_method(f,a, b, tol):

    if f(a)*f(b) > 0:
        print("No root found.")
    else:
        while (b - a) > tol:
            midpoint = a - f(a)*(b-a)/(f(b)-f(a))
            if f(midpoint) == 0:
                return(midpoint)
            elif f(a)*f(midpoint) < 0: 
                b = midpoint
            else:
                a = midpoint
        return(midpoint)
     
sample=10000
execution_time = timeit.timeit(lambda:secant_method(fx,-1, 5, 0.0001), number=10000)
print("Average Time:", execution_time/10000)
print("Result:",secant_method(fx,-1, 5, 0.0001))

